// typescript-schema/dist/lib/equals.js
var m = require('./model');
function createEquals(specificEquals) {
    return function (m) {
        var forElement = this;
        if (forElement.modelKind === m.modelKind) {
            return specificEquals.call(forElement, m);
        }
        else {
            return false;
        }
    };
}
exports.createEquals = createEquals;
function createTypeEquals(specificEquals) {
    return createEquals(function (type) {
        var forType = this;
        if (type.typeKind === forType.typeKind) {
            return specificEquals.call(forType, type);
        }
        else {
            return false;
        }
    });
}
exports.createTypeEquals = createTypeEquals;
function createExpressionEquals(specificEquals) {
    var forExpression = this;
    return createEquals.call(forExpression, function (expression) {
        if (expression.expressionKind === forExpression.expressionKind) {
            return specificEquals.call(forExpression, expression);
        }
        else {
            return false;
        }
    });
}
exports.createExpressionEquals = createExpressionEquals;
exports.typeContainerEquals = createEquals(function (tc) {
    var typeContainer = this;
    if (typeContainer.name === tc.name && typeContainer.typeContainerKind === tc.typeContainerKind) {
        if (tc.typeContainerKind === m.TypeContainerKind.NAMESPACE) {
            return typeContainer.parent.equals(tc.parent);
        }
        else {
            return true;
        }
    }
    else {
        return false;
    }
});
exports.classConstructorEquals = createEquals(function (cc) {
    var classConstuctor = this;
    return classConstuctor.name === cc.name && classConstuctor.parent.equals(cc.parent);
});
exports.interfaceConstructorEquals = createEquals(function (ic) {
    var interfaceConstructor = this;
    return interfaceConstructor.name === ic.name && interfaceConstructor.parent.equals(ic.parent);
});
exports.classEquals = createTypeEquals(function (c) {
    var cls = this;
    if (!cls.typeConstructor.equals(c.typeConstructor)) {
        return false;
    }
    if ((cls.typeArguments && !c.typeArguments) || (!cls.typeArguments && c.typeArguments)) {
        return false;
    }
    if (cls.typeArguments.length !== c.typeArguments.length) {
        return false;
    }
    for (var j = 0; j < cls.typeArguments.length; j++) {
        if (!cls.typeArguments[j].equals(c.typeArguments[j])) {
            return false;
        }
    }
    return true;
});
exports.interfaceEquals = createTypeEquals(function (i) {
    var int = this;
    if (!int.typeConstructor.equals(i.typeConstructor)) {
        return false;
    }
    if ((int.typeArguments && !i.typeArguments) || (!int.typeArguments && i.typeArguments)) {
        return false;
    }
    if (int.typeArguments.length !== i.typeArguments.length) {
        return false;
    }
    for (var j = 0; j < int.typeArguments.length; j++) {
        if (!int.typeArguments[j].equals(i.typeArguments[j])) {
            return false;
        }
    }
    return true;
});
exports.typeParameterEquals = createEquals(function (tp) {
    var typeParameter = this;
    return typeParameter.name === tp.name && typeParameter.parent.equals(tp.parent);
});
exports.compositeTypeEquals = createTypeEquals(function (ct) {
    var compositeType = this;
    var keys = Object.keys(compositeType.members);
    for (var i = 0; i < keys.length; i++) {
        var key = keys[i];
        var thisMember = compositeType.members[key];
        var thatMember = ct.members[key];
        if (!thatMember || !thisMember.type.equals(thatMember.type) || !thisMember.optional === thatMember.optional) {
            return false;
        }
        if (compositeType.index && (!ct.index || !compositeType.index.keyType.equals(ct.index.keyType) || !compositeType.index.valueType.equals(ct.index.valueType))) {
            return false;
        }
        if (compositeType.calls) {
            if (!ct.calls || compositeType.calls.length !== ct.calls.length) {
                return false;
            }
            for (var j = 0; j < compositeType.calls.length; j++) {
                if (!compositeType.calls[j].equals(ct.calls[j])) {
                    return false;
                }
            }
        }
    }
    return true;
});
exports.indexEquals = createEquals(function (i) {
    var index = this;
    return index.parent.equals(i.parent);
});
exports.memberEquals = createEquals(function (m) {
    var member = this;
    return member.parent.equals(m.parent);
});
exports.valueEquals = createEquals(function (v) {
    var value = this;
    return value.name === v.name && value.parent.equals(v.parent);
});
exports.enumMemberEquals = createEquals(function (em) {
    var enumMember = this;
    return enumMember.name === em.name && enumMember.parent.equals(em.parent);
});
exports.enumEquals = createTypeEquals(function (e) {
    var enumType = this;
    return enumType.name === e.name && enumType.parent.equals(e.parent);
});
exports.typeAliasEquals = createTypeEquals(function (a) {
    var typeAlias = this;
    return typeAlias.name === a.name && typeAlias.parent.equals(a.parent);
});
exports.primitiveTypeEquals = createTypeEquals(function (p) {
    var primitiveType = this;
    return primitiveType.primitiveTypeKind === p.primitiveTypeKind;
});
exports.decoratorTypeEquals = function (m) {
    var decoratorType = this;
    if (exports.functionTypeEquals(m)) {
        return decoratorType.decoratorTypeKind === m.decoratorTypeKind;
    }
    else {
        return false;
    }
};
exports.functionTypeEquals = createTypeEquals(function (ft) {
    var functionType = this;
    if (ft.parameters.length !== functionType.parameters.length) {
        return false;
    }
    for (var i = 0; i < ft.parameters.length; i++) {
        var thisParameter = functionType.parameters[i];
        var thatParameter = ft.parameters[i];
        if (thisParameter.name !== thatParameter.name || !thisParameter.type.equals(thatParameter.type) || thisParameter.optional !== thatParameter.optional) {
            return false;
        }
    }
    if ((functionType.type && !ft.type) || (!functionType.type && ft.type)) {
        return false;
    }
    if (functionType.type && !functionType.type.equals(ft.type)) {
        return false;
    }
    if ((functionType.typeParameters && !ft.typeParameters) || (!functionType.typeParameters && ft.typeParameters)) {
        return false;
    }
    if (functionType.typeParameters) {
        if (functionType.typeParameters.length !== ft.typeParameters.length) {
            return false;
        }
        for (var i = 0; i < functionType.typeParameters.length; i++) {
            var thisParameter = functionType.typeParameters[i];
            var thatParameter = ft.typeParameters[i];
            if ((thisParameter.extends && !thatParameter.extends) || (!thisParameter.extends && thatParameter.extends)) {
                return false;
            }
            if (thisParameter.extends && thisParameter.name !== thatParameter.name && !thisParameter.extends.equals(thatParameter.extends)) {
                return false;
            }
        }
    }
    return true;
});
exports.parameterEquals = createEquals(function (p) {
    var parameter = this;
    return parameter.name === p.name && parameter.parent.equals(p.parent);
});
exports.tupleTypeEquals = createTypeEquals(function (t) {
    var tupleType = this;
    if (tupleType.elements.length !== t.elements.length) {
        return false;
    }
    for (var i = 0; i < tupleType.elements.length; i++) {
        if (!tupleType.elements[i].equals(t.elements[i])) {
            return false;
        }
    }
    return true;
});
exports.unionTypeEquals = createTypeEquals(function (t) {
    var unionType = this;
    if (unionType.types.length !== t.types.length) {
        return false;
    }
    for (var i = 0; i < unionType.types.length; i++) {
        if (!unionType.types[i].equals(t.types[i])) {
            return false;
        }
    }
    return true;
});
exports.typeQueryEquals = createTypeEquals(function (tQ) {
    var typeQuery = this;
    return typeQuery.type.equals(tQ.type);
});
exports.literalExpressionEquals = createExpressionEquals(function (rE) {
    var literalExpression = this;
    return literalExpression.value === rE.value;
});
exports.arrayExpressionEquals = createExpressionEquals(function (aE) {
    var arrayExpression = this;
    if (arrayExpression.elements.length !== aE.elements.length) {
        return false;
    }
    for (var i = 0; i < arrayExpression.elements.length; i++) {
        if (!arrayExpression.elements[i].equals(aE.elements[i])) {
            return false;
        }
    }
    return true;
});
exports.objectExpressionEquals = createExpressionEquals(function (oE) {
    var objectExpression = this;
    var keys = Object.keys(objectExpression);
    for (var i = 0; i < keys.length; i++) {
        var thisProp = objectExpression[keys[i]];
        var thatProp = oE[keys[i]];
        if (!thatProp || !thisProp.equals(thatProp)) {
            return false;
        }
    }
    return true;
});
exports.classExpressionEquals = createExpressionEquals(function (cE) {
    var classExpression = this;
    return classExpression.class.equals(cE.class);
});
exports.complexExpressionEquals = createExpressionEquals(function (cE) {
    return undefined;
});
exports.enumExpressionEquals = createExpressionEquals(function (eE) {
    var enumExpression = this;
    return enumExpression.enum.equals(eE.enum) && enumExpression.value === eE.value;
});
exports.decoratorEquals = createEquals(function (d) {
    var decorator = this;
    if (!decorator.parent.equals(d.parent)) {
        return false;
    }
    if (decorator.parameters && !d.parameters || (!decorator.parameters && d.parameters)) {
        return false;
    }
    if (decorator.parameters) {
        if (decorator.parameters.length !== d.parameters.length) {
            return false;
        }
        for (var i = 0; i < decorator.parameters.length; i++) {
            if (!decorator.parameters[i].equals(d.parameters[i])) {
                return false;
            }
        }
    }
    return true;
});
