// typescript-schema/dist/lib/model.js
// TODO: Support ES6 Symbols
(function (TypeKind) {
    TypeKind[TypeKind["PRIMITIVE"] = 1] = "PRIMITIVE";
    TypeKind[TypeKind["ENUM"] = 2] = "ENUM";
    TypeKind[TypeKind["FUNCTION"] = 3] = "FUNCTION";
    TypeKind[TypeKind["TUPLE"] = 4] = "TUPLE";
    TypeKind[TypeKind["UNION"] = 5] = "UNION";
    TypeKind[TypeKind["COMPOSITE"] = 6] = "COMPOSITE";
    TypeKind[TypeKind["INTERFACE"] = 7] = "INTERFACE";
    TypeKind[TypeKind["CLASS"] = 8] = "CLASS";
    TypeKind[TypeKind["TYPE_QUERY"] = 9] = "TYPE_QUERY";
    TypeKind[TypeKind["TYPE_ALIAS"] = 10] = "TYPE_ALIAS";
})(exports.TypeKind || (exports.TypeKind = {}));
var TypeKind = exports.TypeKind;
(function (ValueKind) {
    ValueKind[ValueKind["VAR"] = 1] = "VAR";
    ValueKind[ValueKind["LET"] = 2] = "LET";
    ValueKind[ValueKind["CONST"] = 3] = "CONST";
    ValueKind[ValueKind["FUNCTION"] = 4] = "FUNCTION";
})(exports.ValueKind || (exports.ValueKind = {}));
var ValueKind = exports.ValueKind;
(function (ExpressionKind) {
    ExpressionKind[ExpressionKind["STRING"] = 1] = "STRING";
    ExpressionKind[ExpressionKind["NUMBER"] = 2] = "NUMBER";
    ExpressionKind[ExpressionKind["BOOLEAN"] = 3] = "BOOLEAN";
    ExpressionKind[ExpressionKind["CLASS"] = 4] = "CLASS";
    ExpressionKind[ExpressionKind["OBJECT"] = 5] = "OBJECT";
    ExpressionKind[ExpressionKind["ARRAY"] = 6] = "ARRAY";
    ExpressionKind[ExpressionKind["COMPLEX"] = 7] = "COMPLEX";
    ExpressionKind[ExpressionKind["ENUM"] = 8] = "ENUM";
})(exports.ExpressionKind || (exports.ExpressionKind = {}));
var ExpressionKind = exports.ExpressionKind;
(function (PrimitiveTypeKind) {
    PrimitiveTypeKind[PrimitiveTypeKind["STRING"] = 1] = "STRING";
    PrimitiveTypeKind[PrimitiveTypeKind["BOOLEAN"] = 2] = "BOOLEAN";
    PrimitiveTypeKind[PrimitiveTypeKind["NUMBER"] = 3] = "NUMBER";
    PrimitiveTypeKind[PrimitiveTypeKind["VOID"] = 4] = "VOID";
    PrimitiveTypeKind[PrimitiveTypeKind["ANY"] = 5] = "ANY";
})(exports.PrimitiveTypeKind || (exports.PrimitiveTypeKind = {}));
var PrimitiveTypeKind = exports.PrimitiveTypeKind;
(function (ModelKind) {
    ModelKind[ModelKind["TYPE_CONTAINER"] = 1] = "TYPE_CONTAINER";
    ModelKind[ModelKind["CLASS_CONSTRUCTOR"] = 2] = "CLASS_CONSTRUCTOR";
    ModelKind[ModelKind["INTERFACE_CONSTRUCTOR"] = 3] = "INTERFACE_CONSTRUCTOR";
    ModelKind[ModelKind["TYPE"] = 4] = "TYPE";
    ModelKind[ModelKind["INDEX"] = 5] = "INDEX";
    ModelKind[ModelKind["TYPE_PARAMETER"] = 6] = "TYPE_PARAMETER";
    ModelKind[ModelKind["PARAMETER"] = 7] = "PARAMETER";
    ModelKind[ModelKind["DECORATED_PARAMETER"] = 8] = "DECORATED_PARAMETER";
    ModelKind[ModelKind["MEMBER"] = 9] = "MEMBER";
    ModelKind[ModelKind["DECORATED_MEMBER"] = 10] = "DECORATED_MEMBER";
    ModelKind[ModelKind["DECORATOR"] = 11] = "DECORATOR";
    ModelKind[ModelKind["EXPRESSION"] = 12] = "EXPRESSION";
    ModelKind[ModelKind["VALUE"] = 13] = "VALUE";
    ModelKind[ModelKind["ENUM_MEMBER"] = 14] = "ENUM_MEMBER";
})(exports.ModelKind || (exports.ModelKind = {}));
var ModelKind = exports.ModelKind;
(function (TypeContainerKind) {
    TypeContainerKind[TypeContainerKind["MODULE"] = 1] = "MODULE";
    TypeContainerKind[TypeContainerKind["NAMESPACE"] = 2] = "NAMESPACE";
})(exports.TypeContainerKind || (exports.TypeContainerKind = {}));
var TypeContainerKind = exports.TypeContainerKind;
(function (DecoratorTypeKind) {
    DecoratorTypeKind[DecoratorTypeKind["CLASS"] = 1] = "CLASS";
    DecoratorTypeKind[DecoratorTypeKind["PROPERTY"] = 2] = "PROPERTY";
    DecoratorTypeKind[DecoratorTypeKind["METHOD"] = 3] = "METHOD";
    DecoratorTypeKind[DecoratorTypeKind["PARAMETER"] = 4] = "PARAMETER";
})(exports.DecoratorTypeKind || (exports.DecoratorTypeKind = {}));
var DecoratorTypeKind = exports.DecoratorTypeKind;
exports.STRING = {
    modelKind: ModelKind.TYPE,
    primitiveTypeKind: PrimitiveTypeKind.STRING,
    typeKind: TypeKind.PRIMITIVE,
    equals: function (p) {
        return p.primitiveTypeKind === PrimitiveTypeKind.STRING;
    }
};
exports.BOOLEAN = {
    modelKind: ModelKind.TYPE,
    primitiveTypeKind: PrimitiveTypeKind.BOOLEAN,
    typeKind: TypeKind.PRIMITIVE,
    equals: function (p) {
        return p.primitiveTypeKind === PrimitiveTypeKind.BOOLEAN;
    }
};
exports.NUMBER = {
    modelKind: ModelKind.TYPE,
    primitiveTypeKind: PrimitiveTypeKind.NUMBER,
    typeKind: TypeKind.PRIMITIVE,
    equals: function (p) {
        return p.primitiveTypeKind === PrimitiveTypeKind.NUMBER;
    }
};
exports.ANY = {
    modelKind: ModelKind.TYPE,
    primitiveTypeKind: PrimitiveTypeKind.ANY,
    typeKind: TypeKind.PRIMITIVE,
    equals: function (p) {
        return p.primitiveTypeKind === PrimitiveTypeKind.ANY;
    }
};
exports.VOID = {
    modelKind: ModelKind.TYPE,
    primitiveTypeKind: PrimitiveTypeKind.VOID,
    typeKind: TypeKind.PRIMITIVE,
    equals: function (p) {
        return p.primitiveTypeKind === PrimitiveTypeKind.VOID;
    }
};
