// typescript-schema/dist/lib/serialiserParser.js
var m = require('./model');
function stringifyModuleReplacer(key, value) {
    if (key === 'typeKind') {
        switch (value) {
            case m.TypeKind.PRIMITIVE:
                return 'PRIMITIVE';
            case m.TypeKind.ENUM:
                return 'ENUM';
            case m.TypeKind.TYPE_QUERY:
                return 'TYPE_QUERY';
            case m.TypeKind.FUNCTION:
                return 'FUNCTION';
            case m.TypeKind.TUPLE:
                return 'TUPLE';
            case m.TypeKind.UNION:
                return 'UNION';
            case m.TypeKind.COMPOSITE:
                return 'COMPOSITE';
            case m.TypeKind.INTERFACE:
                return 'INTERFACE';
            case m.TypeKind.CLASS:
                return 'CLASS';
        }
        return undefined;
    }
    else {
        return value;
    }
}
exports.stringifyModuleReplacer = stringifyModuleReplacer;
function parseModuleReplacer(key, value) {
    if (key === 'typeKind') {
        switch (value.toUpperCase()) {
            case 'PRIMITIVE':
                return m.TypeKind.PRIMITIVE;
            case 'ENUM':
                return m.TypeKind.ENUM;
            case 'TYPE_QUERY':
                return m.TypeKind.TYPE_QUERY;
            case 'FUNCTION':
                return m.TypeKind.FUNCTION;
            case 'TUPLE':
                return m.TypeKind.TUPLE;
            case 'UNION':
                return m.TypeKind.UNION;
            case 'COMPOSITE':
                return m.TypeKind.COMPOSITE;
            case 'INTERFACE':
                return m.TypeKind.INTERFACE;
            case 'CLASS':
                return m.TypeKind.CLASS;
            default:
                throw new Error('Unrecognised typeKind value: ' + value);
        }
    }
    else {
        return value;
    }
}
exports.parseModuleReplacer = parseModuleReplacer;
function stringifyModules(mods) {
    return JSON.stringify(mods, stringifyModuleReplacer, '  ');
}
exports.stringifyModules = stringifyModules;
function stringifyModule(mod) {
    return JSON.stringify(mod, stringifyModuleReplacer, '  ');
}
exports.stringifyModule = stringifyModule;
function parseModules(modsString) {
    return JSON.parse(modsString, parseModuleReplacer);
}
exports.parseModules = parseModules;
function parseModule(modString) {
    return JSON.parse(modString, parseModuleReplacer);
}
exports.parseModule = parseModule;
