// typescript-schema/dist/lib/toString.js
var m = require('./model');
function typeToString(type) {
    if (type.typeKind) {
        switch (type.typeKind) {
            case m.TypeKind.FUNCTION:
                return functionTypeToString(type);
            case m.TypeKind.CLASS:
                return classToString(type);
            case m.TypeKind.INTERFACE:
                return interfaceToString(type);
            case m.TypeKind.TUPLE:
                return tupleTypeToString(type);
            case m.TypeKind.UNION:
                return unionTypeToString(type);
            case m.TypeKind.PRIMITIVE:
                return primitiveTypeToString(type);
        }
    }
    else {
        var typeParameter = type;
        return '@' + typeParameter.name;
    }
}
exports.typeToString = typeToString;
function typeContainerToString(typeContainer) {
    var str = typeContainer.name;
    while (typeContainer.typeContainerKind === m.TypeContainerKind.NAMESPACE) {
        typeContainer = typeContainer.parent;
        if (typeContainer.name !== '') {
            str = typeContainer.name + ':' + str;
        }
    }
    return str;
}
exports.typeContainerToString = typeContainerToString;
function classConstructorToString(classConstructor) {
    return typeContainerToString(classConstructor.parent) + ':' + classConstructor.name;
}
exports.classConstructorToString = classConstructorToString;
function interfaceConstructorToString(interfaceConstructor) {
    return typeContainerToString(interfaceConstructor.parent) + ':' + interfaceConstructor.name;
}
exports.interfaceConstructorToString = interfaceConstructorToString;
function primitiveTypeToString(type) {
    switch (type.primitiveTypeKind) {
        case m.PrimitiveTypeKind.STRING:
            return 'string';
        case m.PrimitiveTypeKind.BOOLEAN:
            return 'boolean';
        case m.PrimitiveTypeKind.NUMBER:
            return 'number';
        case m.PrimitiveTypeKind.ANY:
            return 'any';
    }
}
exports.primitiveTypeToString = primitiveTypeToString;
function functionTypeToString(type) {
    var str = '(';
    if (type.parameters.length > 0) {
        str += type.parameters[0].name + ":" + typeToString(type.parameters[0].type);
        for (var i = 1; i < type.parameters.length; i++) {
            str += "," + type.parameters[0].name + ":" + typeToString(type.parameters[0].type);
        }
    }
    return str + (")=>(" + (type.type ? typeToString(type.type) : 'void') + ")");
}
exports.functionTypeToString = functionTypeToString;
function tupleTypeToString(type) {
    var str = '[';
    str += typeToString(type.elements[0]);
    for (var i = 1; i < type.elements.length; i++) {
        str += ',' + typeToString(type.elements[i]);
    }
    return str + ']';
}
exports.tupleTypeToString = tupleTypeToString;
function unionTypeToString(type) {
    var str = typeToString(type.types[0]);
    for (var i = 1; type.types.length; i++) {
        str += '|' + typeToString(type.types[i]);
    }
    return str;
}
exports.unionTypeToString = unionTypeToString;
function typeArgsToString(typeArgs) {
    if (typeArgs && typeArgs.length > 0) {
        var str = '<';
        str += typeToString(typeArgs[0]);
        for (var i = 1; i < typeArgs.length; i++) {
            str += ',' + typeToString(typeArgs[i]);
        }
        return str + '>';
    }
    else {
        return '';
    }
}
function classToString(type) {
    var str = classConstructorToString(type.typeConstructor);
    str += typeArgsToString(type.typeArguments);
    return str;
}
exports.classToString = classToString;
function interfaceToString(type) {
    var str = interfaceConstructorToString(type.typeConstructor);
    str += typeArgsToString(type.typeArguments);
    return str;
}
exports.interfaceToString = interfaceToString;
