// typescript-schema/dist/lib/typeCloser.js
var m = require('./model');
var e = require('./equals');
var s = require('./toString');
function closeClassConstructor(classConstructor, typeArgs, parentTypeArgs, closedTypes, cls) {
    return processClass(classConstructor, typeArgs, parentTypeArgs, closedTypes, classConstructor.instanceType, classConstructor.staticType, classConstructor.extends, classConstructor.implements, classConstructor.decorators, cls);
}
exports.closeClassConstructor = closeClassConstructor;
function processClass(classConstructor, typeArgs, parentTypeArgs, closedTypes, instanceType, staticType, clsExtends, clsImplements, decorators, cls) {
    cls = cls || {};
    cls.modelKind = m.ModelKind.TYPE;
    cls.name = classConstructor.name;
    cls.typeKind = m.TypeKind.CLASS;
    cls.typeConstructor = classConstructor;
    cls.constructorParent = classConstructor.parent;
    cls.typeArguments = typeArgs;
    cls.equals = e.classEquals;
    var fqName = s.classConstructorToString(classConstructor);
    var previouslyClosed = getPreviouslyClosed(fqName, closedTypes, typeArgs, parentTypeArgs);
    if (previouslyClosed) {
        cls.instanceType = copyCompositeType(previouslyClosed.instanceType, cls);
        cls.staticType = copyCompositeType(previouslyClosed.staticType, cls);
        cls.extends = previouslyClosed.extends;
        cls.implements = previouslyClosed.implements;
        copyDecorators(previouslyClosed.decorators, cls);
    }
    else {
        var allArgs = getAllArgs(classConstructor.typeParameters, typeArgs, parentTypeArgs);
        closedTypes = closedTypes || {};
        cls.instanceType = processCompositeType(allArgs, instanceType, true);
        cls.staticType = processCompositeType(allArgs, staticType, true);
        cls.instanceType.parent = cls;
        cls.staticType.parent = cls;
        addClosed(fqName, closedTypes, cls, parentTypeArgs);
        if (decorators) {
            cls.decorators = decorators.map(function (decorator) {
                return {
                    modelKind: m.ModelKind.DECORATOR,
                    parent: cls,
                    equals: e.decoratorEquals,
                    decoratorType: decorator.decoratorType,
                    parameters: decorator.parameters
                };
            });
        }
        if (clsExtends) {
            if (!clsExtends.typeKind) {
                if (clsExtends.modelKind) {
                    var cc = clsExtends;
                    var typeArgs_1 = (cc).typeParameters ? (cc).typeParameters.map(function () {
                        return m.ANY;
                    }) : [];
                    clsExtends = closeClassConstructor(cc, typeArgs_1, parentTypeArgs, closedTypes);
                }
                else {
                    clsExtends = clsExtends();
                }
            }
            cls.extends = processClass(clsExtends.typeConstructor, clsExtends.typeArguments, allArgs, closedTypes, clsExtends.instanceType, clsExtends.staticType, clsExtends.extends, clsExtends.implements, clsExtends.decorators);
        }
        if (clsImplements) {
            cls.implements = [];
            cls.implements = clsImplements.map(function (impl) {
                if (!impl.typeKind) {
                    if (impl.modelKind) {
                        var ic = impl;
                        var typeArgs_2 = (ic).typeParameters ? (ic).typeParameters.map(function () {
                            return m.ANY;
                        }) : [];
                        impl = closeInterfaceConstructor(ic, typeArgs_2, parentTypeArgs, closedTypes);
                    }
                    else {
                        impl = impl();
                    }
                }
                return processInterface(impl.typeConstructor, impl.typeArguments, allArgs, closedTypes, impl.instanceType, impl.extends);
            });
        }
    }
    return cls;
}
exports.processClass = processClass;
function closeInterfaceConstructor(interfaceConstructor, typeArgs, parentTypeArgs, closedTypes, int) {
    return processInterface(interfaceConstructor, typeArgs, parentTypeArgs, closedTypes, interfaceConstructor.instanceType, interfaceConstructor.extends, int);
}
exports.closeInterfaceConstructor = closeInterfaceConstructor;
function processInterface(interfaceConstructor, typeArgs, parentTypeArgs, closedTypes, instanceCompositeType, intExtends, int) {
    int = int || {};
    int.modelKind = m.ModelKind.TYPE;
    int.name = interfaceConstructor.name;
    int.typeKind = m.TypeKind.INTERFACE;
    int.typeConstructor = interfaceConstructor;
    int.constructorParent = interfaceConstructor.parent;
    int.typeArguments = typeArgs;
    int.equals = e.interfaceEquals;
    var fqName = s.interfaceConstructorToString(interfaceConstructor);
    var previouslyClosed = getPreviouslyClosed(fqName, closedTypes, typeArgs, parentTypeArgs);
    if (previouslyClosed) {
        int.instanceType = copyCompositeType(previouslyClosed.instanceType, int);
        int.extends = previouslyClosed.extends;
    }
    else {
        var allArgs = getAllArgs(interfaceConstructor.typeParameters, typeArgs, parentTypeArgs);
        closedTypes = closedTypes || {};
        int.instanceType = processCompositeType(allArgs, interfaceConstructor.instanceType, false);
        int.instanceType.parent = int;
        addClosed(fqName, closedTypes, int, parentTypeArgs);
        if (intExtends) {
            int.extends = [];
            int.extends = intExtends.map(function (extend) {
                if (!extend.typeKind) {
                    if (extend.modelKind) {
                        var ic = extend;
                        var typeArgs_3 = (ic).typeParameters ? (ic).typeParameters.map(function () {
                            return m.ANY;
                        }) : [];
                        extend = closeInterfaceConstructor(ic, typeArgs_3, parentTypeArgs, closedTypes);
                    }
                    else {
                        extend = extend();
                    }
                }
                return processInterface(extend.typeConstructor, extend.typeArguments, allArgs, closedTypes, extend.instanceType, extend.extends);
            });
        }
    }
    return int;
}
function copyDecorators(decorators, parent) {
    if (decorators) {
        parent.decorators = decorators.map(function (decorator) {
            return {
                parent: parent,
                decoratorType: decorator.decoratorType,
                parameters: decorator.parameters,
                modelKind: m.ModelKind.DECORATOR,
                equals: e.decoratorEquals
            };
        });
    }
}
function copyCompositeType(ct, parent) {
    var copy = {
        members: {},
        modelKind: m.ModelKind.TYPE,
        typeKind: m.TypeKind.COMPOSITE,
        equals: e.compositeTypeEquals
    };
    Object.keys(ct.members).forEach(function (name) {
        var member = ct.members[name];
        copy.members[name] = {
            parent: copy,
            name: name,
            type: member.type,
            optional: member.optional,
            initialiser: member.initialiser,
            modelKind: member.modelKind,
            equals: e.memberEquals
        };
        copyDecorators(member.decorators, copy.members[name]);
    });
    if (ct.index) {
        copy.index = {
            parent: copy,
            keyType: ct.index.keyType,
            valueType: ct.index.valueType,
            modelKind: m.ModelKind.INDEX,
            equals: e.indexEquals
        };
    }
    if (ct.calls) {
        copy.calls = ct.calls.map(function (call) {
            return call;
        });
    }
    copy.parent = parent;
    return copy;
}
function addClosed(fqName, closedTypes, closedInstance, parentTypeArgs) {
    var closedTypesForName = closedTypes[fqName];
    if (!closedTypesForName) {
        closedTypesForName = [];
        closedTypes[fqName] = closedTypesForName;
    }
    closedTypesForName.push([closedInstance, parentTypeArgs]);
}
function getPreviouslyClosed(fqName, closedTypes, typeArgs, parentTypeArgs) {
    if (closedTypes[fqName]) {
        var closedTypesForName = closedTypes[fqName];
        for (var i = 0; i < closedTypesForName.length; i++) {
            var closedType = closedTypesForName[i];
            var closedTypeArgs = closedType[0].typeArguments;
            var isMatch = true;
            for (var j = 0; j < closedTypeArgs.length; j++) {
                if (!closedTypeArgs[j].equals(typeArgs[j])) {
                    isMatch = false;
                    break;
                }
            }
            if (isMatch) {
                if ((closedType[1] && !parentTypeArgs) || (!closedType[1] && parentTypeArgs)) {
                    isMatch = false;
                }
                else {
                    var closedParentTypeArgs = closedType[1];
                    var keys = Object.keys(closedParentTypeArgs);
                    for (var j = 0; j < keys.length; j++) {
                        var key = keys[j];
                        if (!parentTypeArgs[key]) {
                            isMatch = false;
                            break;
                        }
                        else if (!closedParentTypeArgs[key].equals(parentTypeArgs[key])) {
                            isMatch = false;
                            break;
                        }
                    }
                }
                if (isMatch) {
                    return closedType[0];
                }
            }
        }
    }
}
function substituteType(allArgs, type) {
    if (!type.modelKind) {
        type = type();
    }
    if (type.typeKind) {
        return type;
    }
    else {
        var p = type;
        return allArgs[p.name] || p;
    }
}
function processCompositeType(allArgs, oldCompositeType, isDecorated) {
    var newCompositeType = {
        modelKind: m.ModelKind.TYPE,
        members: {},
        equals: e.compositeTypeEquals,
        typeKind: m.TypeKind.COMPOSITE
    };
    Object.keys(oldCompositeType.members).forEach(function (name) {
        var member = oldCompositeType.members[name];
        newCompositeType.members[name] = {
            modelKind: m.ModelKind.MEMBER,
            parent: newCompositeType,
            equals: e.compositeTypeEquals,
            name: name,
            type: substituteType(allArgs, member.type),
            optional: member.optional
        };
        if (isDecorated) {
            newCompositeType.members[name].decorators = member.decorators;
        }
    });
    if (oldCompositeType.index) {
        newCompositeType.index = {
            modelKind: m.ModelKind.INDEX,
            parent: newCompositeType,
            equals: e.compositeTypeEquals,
            keyType: oldCompositeType.index.keyType,
            valueType: substituteType(allArgs, oldCompositeType.index.valueType)
        };
    }
    if (oldCompositeType.calls) {
        newCompositeType.calls = oldCompositeType.calls.map(function (call) {
            var newCall = {
                modelKind: m.ModelKind.TYPE,
                typeKind: m.TypeKind.FUNCTION,
                equals: e.functionTypeEquals,
                typeParameters: call.typeParameters,
                parameters: call.parameters.map(function (parameter) {
                    var newParameter = {
                        modelKind: parameter.decorators ? m.ModelKind.DECORATED_PARAMETER : m.ModelKind.PARAMETER,
                        parent: newCall,
                        equals: e.parameterEquals,
                        name: parameter.name,
                        type: substituteType(allArgs, parameter.type)
                    };
                    if (isDecorated && parameter.decorators) {
                        newParameter.decorators = parameter.decorators;
                    }
                    return newParameter;
                })
            };
            if (call.type) {
                newCall.type = substituteType(allArgs, call.type);
            }
            return newCall;
        });
    }
    return newCompositeType;
}
function getAllArgs(typeParams, typeArgs, parentTypeArgs) {
    var names = parentTypeArgs ? Object.keys(parentTypeArgs) : null;
    var allArgs = {};
    if (parentTypeArgs) {
        names.forEach(function (name) {
            allArgs[name] = parentTypeArgs[name];
        });
    }
    if (typeParams) {
        typeParams.forEach(function (typeParam, i) {
            allArgs[typeParam.name] = typeArgs[i];
        });
    }
    return allArgs;
}
