// typescript-schema/dist/lib/typeUtils.js
var m = require('./model');
var ROOT = {
    modelKind: m.ModelKind.TYPE_CONTAINER,
    typeContainerKind: m.TypeContainerKind.MODULE,
    name: ''
};
var FUNCTION = {
    modelKind: m.ModelKind.TYPE,
    typeKind: m.TypeKind.INTERFACE,
    name: 'Function',
    constructorParent: ROOT,
    typeConstructor: {
        name: 'Function',
        parent: ROOT
    }
};
function isFunctionVoid(f) {
    return !f.type || f.type.primitiveTypeKind === m.PrimitiveTypeKind.VOID;
}
exports.isFunctionVoid = isFunctionVoid;
function isSubType(potentialSubType, potentialSuperType) {
    if (potentialSubType.typeKind === m.TypeKind.CLASS) {
        var cls = potentialSubType;
        if (potentialSuperType.typeKind === m.TypeKind.CLASS) {
            if (potentialSubType.equals(potentialSuperType)) {
                return true;
            }
            if (cls.extends) {
                return isSubType(cls.extends, potentialSuperType);
            }
            else {
                return false;
            }
        }
        else {
            if (cls.implements) {
                for (var i = 0; i < cls.implements.length; i++) {
                    if (isSubType(cls.implements[i], potentialSuperType)) {
                        return true;
                    }
                }
                return false;
            }
            else {
                return false;
            }
        }
    }
    else {
        var int = potentialSubType;
        if (potentialSuperType.typeKind === m.TypeKind.CLASS) {
            return false;
        }
        else {
            if (potentialSubType.equals(potentialSuperType)) {
                return true;
            }
            if (int.extends) {
                for (var i = 0; i < int.extends.length; i++) {
                    if (isSubType(int.extends[i], potentialSuperType)) {
                        return true;
                    }
                }
                return false;
            }
            else {
                return false;
            }
        }
    }
}
exports.isSubType = isSubType;
function isFunctionType(type) {
    if (type.typeKind === m.TypeKind.FUNCTION) {
        return true;
    }
    else if (type.typeKind === m.TypeKind.INTERFACE) {
        var int = type;
        return isSubType(int, FUNCTION);
        return false;
    }
    return type.typeKind === m.TypeKind.FUNCTION || (type.name === 'Function' && type.constructorParent.name === '');
}
exports.isFunctionType = isFunctionType;
